### PR TITLE
Add SSL connection support

### PR DESCRIPTION
**Pull request by pinepain**
This PR:
 
 - add SSL connection support.

Now it is possible to pass cert files to establish SSL connection. CA cert file is mandatory, when path to it passed via connection constructor credentials array, then `AMQPConnection::connect()` will try to establish SSL connection. Note, that playing with self-generate certs setting pear verification to false may be a good idea. To play a bit with SSL connection see [TLS Support](https://www.rabbitmq.com/ssl.html) RabbitMQ manual and check this awesome [TLS/SSL Certificate Generator](https://github.com/michaelklishin/tls-gen). By default, is you will use vagrant environment that ships within this repo, SSL should certs are already generated and RabbitMQ is configured. See [tests/amqpconnection_ssl.phpt](https://github.com/pinepain/php-amqp/blob/7005beb658959879417f7b437978c6b1063b5e1b/tests/amqpconnection_ssl.phpt) test file for example.

This changeset doesn't break BC but it extend existent API usage.

### CODE REVIEW

File: amqp.c [hunk 1]
```diff
@@ -395,6 +395,10 @@ PHP_INI_BEGIN()
 	PHP_INI_ENTRY("amqp.channel_max",		DEFAULT_CHANNEL_MAX,		PHP_INI_ALL, NULL)
 	PHP_INI_ENTRY("amqp.frame_max",			DEFAULT_FRAME_MAX,			PHP_INI_ALL, NULL)
 	PHP_INI_ENTRY("amqp.heartbeat",			DEFAULT_HEARTBEAT,			PHP_INI_ALL, NULL)
+	PHP_INI_ENTRY("amqp.cacert",			DEFAULT_CACERT,				PHP_INI_ALL, NULL)
+	PHP_INI_ENTRY("amqp.cert",				DEFAULT_CERT,				PHP_INI_ALL, NULL)
+	PHP_INI_ENTRY("amqp.key",				DEFAULT_KEY,				PHP_INI_ALL, NULL)
+	PHP_INI_ENTRY("amqp.verify",			DEFAULT_VERIFY,				PHP_INI_ALL, NULL)
```
Comment (lstrojny) [inline amqp.c]: Does that verify hostnames and peer names?

**Comment by lstrojny** [comment]
Awesome, thank you! I try to get a view more people having a look but great work already.

File: amqp.c (hunk 1 above)
Comment (zaq178miami) [inline amqp.c]: Yes. Before rabbitmq-c 0.8.0 there was no way to verify them separately (see [this changes below](https://github.com/pdezwart/php-amqp/pull/205/files#diff-89309d5b78d7c1a44970257f5ec91eb6R325)), so I make single general option to not bother users with rabbitmq-c API changes. Anyway, in prod all verifications should be turned on.

Comment (pinepain) [inline amqp.c]: Yes. Before rabbitmq-c 0.8.0 there was no way to verify them separately (see this changes below), so I make single general option to not bother users with rabbitmq-c API changes. Anyway, in prod all verifications should be turned on.

**Comment by ZhaoVitapublic** [comment]
Great work! We have use this with librabbitmq-c 0.7.1 on our testing-system, It works well for us!

**Comment by basz** [comment]
Hi, I'm trying to get SSL working rabbitmq so I stumbled on this port. To try it I have build the extension from source and it now accepts those parameters.

Being new to rabbitmq I needed to jump some hoops. Updated erlang and rabbitmq to the latest versions. Received an certificate from let encrypt and guessing the exact configuration values.

    openssl s_server -accept 8444 -cert /Volumes/rabbitmq/ssl/cert.pem  -key /Volumes/rabbitmq/ssl/key.pem    -CAfile /Volumes/rabbitmq/ssl/cacert.pem 

and

    openssl s_client -connect localhost:8444  -CAfile /Volumes/rabbitmq/ssl/cacert.pem 

Apparently I can now create an encrypted connection via the cli, so the certificates seem valid. However as soon as I add the `cacert` option this PR provides an error is return `Socket error: could not connect to host.`. Previously (the released extension, no cacert option) the error was `Library error: a socket error occurred - Potential login failure.`.

ps. I'm using https://curl.haxx.se/ca/cacert.pem as cacert file.

So, my question... Am I on the right track to configuring SSL? Confusion is increased as I see references to client certificates. I don't want/need that if possible because I will need to expose a web-stomp listener and those clients won't have client certificates. I think client certificates are used to authenticate to the clients with server, but I would like to simply use the username/passwords for that. That's what I understand of it at least.

I apologise the off topic help request, but info is sparse on this one...

**Comment by pinepain** [comment]
Have a look into [test file](https://github.com/pdezwart/php-amqp/pull/205/files#diff-58fcf0dbccb856a29cbcd403737a92dbR9) for examples. Do you have hosts verify on or off? Can you post some code you use to connect? And please, specify PHP version you are using (less likely it cause any problem, just to know on which version this changes are tested).

**Comment by basz** [comment]
$ php -v
    PHP 7.0.4 (cli) (built: Mar  4 2016 13:15:04) ( NTS )

more info: https://gist.github.com/basz/1e24610d5fae9138c4b49ea2eaaf75cf

this bothers me : Verify return code: 21 (unable to verify the first certificate)

```
php ./amqp.php 
Socket error: could not connect to host.#0 /Users/bas/Documents/Projects/PLHW.ProjectHealthyFeet/applications/hf-app-api/amqp.php(32): AMQPConnection->connect()
#1 /Users/bas/Documents/Projects/PLHW.ProjectHealthyFeet/applications/hf-app-api/amqp.php(41): {closure}()
#2 {main}Socket error: could not connect to host.#0 /Users/bas/Documents/Projects/PLHW.ProjectHealthyFeet/applications/hf-app-api/amqp.php(32): AMQPConnection->connect()
```

**Comment by basz** [comment]
oh yes. i figured it out. What needs to be done is to use this certificate as cacert

wget -q -O - "$@" https://letsencrypt.org/certs/lets-encrypt-x3-cross-signed.pem.txt > /etc/rabbitmq/ssl/cacert.pem

awesome thanks

**Comment by pinepain** [comment]
ping @lstrojny what our roadmap on this PR?

**Comment by basz** [comment]
I believe it's just a matter of merging/releasing. Works fine so far.

**Comment by ZhaoVitapublic** [comment]
It works fine on our testing system since few weeks. I hope, the ssl support can be going to merging/release in further.

**Comment by lstrojny** [comment]
We are going to tag the 1.7.0 as stable, merge the new features and roll an alpha/beta

**Comment by lstrojny** [comment]
@pinepain 1.7.0 has been released. 1.8.0-dev lives in master, 1.7.1-dev lives in release/1.7.

**Comment by lstrojny** [comment]
@pinepain can you rebase against master so we can merge this?

**Comment by pinepain** [comment]
@lstrojny rebased. I'll review it one more time until travis running builds.

**Comment by lstrojny** [comment]
Manually merged

**Comment by lstrojny** [comment]
As always, great work @pinepain, thank you!

**Comment by basz** [comment]
Thanks! Great addition.

In what version did this end up now?

**Comment by lstrojny** [comment]
@basz right now it lives in master which is going to be the upcoming 1.8 release.

**Comment by ZhaoVitapublic** [comment]
Thanks all! 
Is really cool to have this feature in the extension.

**Pull request merged**